sycl/source/detail/device_filter.cpp

```diff
@@ -0,0 +1,115 @@
+//==------------------- device_filter.cpp ----------------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <CL/sycl/detail/device_filter.hpp>
+#include <CL/sycl/info/info_desc.hpp>
+#include <detail/config.hpp>
+#include <detail/device_impl.hpp>
+
+#include <cstring>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+
+device_filter::device_filter(const std::string &FilterString) {
+  const std::array<std::pair<std::string, info::device_type>, 5>
+      SyclDeviceTypeMap = {{{"host", info::device_type::host},
+                            {"cpu", info::device_type::cpu},
+                            {"gpu", info::device_type::gpu},
+                            {"acc", info::device_type::accelerator},
+                            {"*", info::device_type::all}}};
+  const std::array<std::pair<std::string, backend>, 4> SyclBeMap = {
+      {{"opencl", backend::opencl},
+       {"level_zero", backend::level_zero},
+       {"cuda", backend::cuda},
+       {"*", backend::all}}};
+
+  // handle the optional 1st entry, backend
+  size_t Cursor = 0;
+  size_t ColonPos = FilterString.find(":", Cursor);
+  auto It = std::find_if(
```

Comment:
Please comment the code from time to time.
Just a test: ask you N+1/N+2 manager to explain the code... :-)

Reply:
Thanks. Done.